extensions/amp-analytics/0.1/visibility-helper.js

```diff
@@ -0,0 +1,397 @@
+ * limitations under the License.
+ */
+
+import {dev} from '../../../src/log';
+
+
+/**
+ * A helper class that implements visibility calculations based on the
+ * visibility ratio. It's used for documents, embeds and individual element.
+ * @implements {../../../src/service.Disposable}
+ */
+export class VisibilityHelper {
+  /**
+   * @param {?VisibilityHelper} parent
+   * @param {!Object<string, *>} spec
+   * @param {number=} opt_iniVisibility
+   * @param {boolean=} opt_factorParent
+   */
+  constructor(parent, spec, opt_iniVisibility, opt_factorParent) {
+    /** @const @private */
+    this.parent_ = parent;
+
+    /**
+     * Whether this visibility is in the intersection with parent. Thus the
+     * final visibility will be this visibility times parent.
+     * @const @private {boolean}
+     */
+    this.factorParent_ = opt_factorParent || false;
+
+    /** @const @private */
+    this.eventPromise_ = new Promise(resolve => {
+      /** @private {?function()} */
+      this.eventResolver_ = resolve;
+    });
+
+    /** @private {?Array<!VisibilityHelper>} */
+    this.children_ = null;
+
+    /** @private {!Array<!UnsubscribeDef>} */
+    this.unsubscribe_ = [];
+
+    // Spec.
+
+    /** @const @private {number} */
+    this.visiblePercentageMin_ =
+        Number(spec['visiblePercentageMin']) / 100 || 0;
+
+    /** @const @private {number} */
+    this.visiblePercentageMax_ =
+        Number(spec['visiblePercentageMax']) / 100 || 1;
+
+    /** @const @private {time} */
+    this.totalTimeMin_ = Number(spec['totalTimeMin']) || 0;
+
+    /** @const @private {time} */
+    this.totalTimeMax_ = Number(spec['totalTimeMax']) || Infinity;
+
+    /** @const @private {time} */
+    this.continuousTimeMin_ = Number(spec['continuousTimeMin']) || 0;
+
+    /** @const @private {time} */
+    this.continuousTimeMax_ = Number(spec['continuousTimeMax']) || Infinity;
+
+    // Runtime.
+
+    /** @const @private {time} */
+    this.createdTime_ = Date.now();
+
+    /** @private {number} */
+    this.ownVisibility_ = opt_iniVisibility || 0;
+
+    /** @private {boolean} */
+    this.blocked_ = false;
+
+    /** @private {?number} */
+    this.scheduledRunId_ = null;
+
+    /** @private {boolean} */
+    this.matchesVisibility_ = false;
+
+    /** @private {time} */
+    this.continuousTime_ = 0;
+
+    /** @private {time} */
+    this.maxContinuousVisibleTime_ = 0;
+
+    /** @private {time} */
+    this.totalVisibleTime_ = 0;
+
+    /** @private {time} */
+    this.firstSeenTime_ = 0;
+
+    /** @private {time} */
+    this.lastSeenTime_ = 0;
+
+    /** @private {time} */
+    this.fistVisibleTime_ = 0;
+
+    /** @private {time} */
+    this.lastVisibleTime_ = 0;
+
+    /** @private {time} */
+    this.loadTimeVisibility_ = 0;
+
+    /** @private {number} */
+    this.minVisiblePercentage_ = 0;
+
+    /** @private {number} */
+    this.maxVisiblePercentage_ = 0;
+
+    /** @private {time} */
+    this.lastVisibleUpdateTime_ = 0;
```

Comment:
I thought about it. But I'd rather avoid extra object creation and extra object lookup in this case. It's slightly inconvenient, but has more benefits this way. Also, the difference between internal state and a metric is extremely small. If you'd like, I can create a separate object for config as a tradeoff. Then, we'll have a single config and everything else will be an internal state?